### PR TITLE
Clean-up when destroying environment

### PR DESCRIPTION
**Pull request by MattiasGees**
**What this PR does / why we need it**:
This will destroy a complete environment and all of its clusters:
- Run terraform destroy
- Remove state files from S3 and dynamodb
- Move environment folder to .archive
- Remove environment config from tarmak.yaml

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
part of #523 
**Special notes for your reviewer**:

**Release note**:

```release-note
Add terraform environment destroy command
```

### CODE REVIEW

File: pkg/tarmak/terraform.go [hunk 1]
```diff
@@ -179,6 +179,13 @@ func (c *CmdTerraform) Destroy() error {
 		return err
 	}
 
+	c.tarmak.cluster.Log().Info("Removing environment")
+
+	err = c.tarmak.environment.Remove()
```
Comment (MattiasGees) [inline pkg/tarmak/terraform.go]: Not sure if we have to make this command part of `tarmak cluster destroy`. Input is welcome for this one!

Comment (MattiasGees) [reply pkg/tarmak/terraform.go]: ok

**Comment by MattiasGees** [comment]
Talked about this with the team. Need to make a few changes.
Change `--force` to `--auto-approve=false`. Together with this, we can use user interaction to run the destroy.

**Comment by MattiasGees** [comment]
All functionality works. UI interaction is implemented. 

One task is remaining. When de skip the cluster destroy part, we need to make sure nothing is running on AWS anymore. Only after that we can delete the the config from tarmak.yaml and move the environment folder to `.archive`

**Comment by MattiasGees** [comment]
This is ready for review
/unassign
/assign @simonswine

File: pkg/terraform/terraform.go [hunk 2]
```diff
@@ -21,7 +22,6 @@ import (
 
 	"github.com/jetstack/tarmak/pkg/tarmak/interfaces"
 	"github.com/jetstack/tarmak/pkg/tarmak/utils"
-	"github.com/jetstack/tarmak/pkg/terraform/providers/tarmak/rpc"
```
Comment (simonswine) [inline pkg/terraform/terraform.go]: can we keep it like that?

File: cmd/tarmak/cmd/environment.go [hunk 3]
```diff
@@ -11,6 +12,25 @@ var environmentCmd = &cobra.Command{
...
+		"auto-approve destroy of a complete environment",
+	)
+
+	fs.StringVarP(
+		&store.Name,
+		"name",
+		"n",
+		"",
+		"name of the environment to destroy",
+	)
```
Comment (simonswine) [inline cmd/tarmak/cmd/environment.go]: I would prefer to have that as an argument rather than a flag

This aligns closer with `clusters set-current test` and other commands

Comment (JoshVanL) [reply cmd/tarmak/cmd/environment.go]: I agree

File: pkg/tarmak/config/config.go [hunk 4]
```diff
@@ -281,6 +281,30 @@ func (c *Config) AppendEnvironment(env *tarmakv1alpha1.Environment) error {
...
 }
 
+func (c *Config) RemoveEnvironment(environment string) error {
+	for key, env := range c.conf.Environments {
+		if env.Name == environment {
+			c.conf.Environments = append(c.conf.Environments[:key], c.conf.Environments[key+1:]...)
+		}
+	}
+
+	done := false
```
Comment (JoshVanL) [inline pkg/tarmak/config/config.go]: This could be a bit nicer, something like 
```
	for i := 0; i < len(c.conf.Clusters); i++ {
		if c.conf.Clusters[i].Environment == environment {
			c.conf.Clusters = append(c.conf.Clusters[:i], c.conf.Clusters[i+1:]...)
			i -= 1
		}
	}
```

File: pkg/tarmak/provider/amazon/tf_remote_state.go [hunk 5]
```diff
@@ -385,3 +385,153 @@ func (a *Amazon) initRemoteStateKMS() error {
...
+
+	_, err = svc.DeleteBucket(&s3.DeleteBucketInput{
+		Bucket: aws.String(a.RemoteStateName()),
+	})
+
+	if err != nil {
+		ec2err, ok := err.(awserr.Error)
+		if ok && ec2err.Code() == "BucketNotEmpty" {
+			resp, err := svc.ListObjectVersions(
+				&s3.ListObjectVersionsInput{
```
Comment (JoshVanL) [inline pkg/tarmak/provider/amazon/tf_remote_state.go]: Maybe this could go into a separate function, lots of tabs :grimacing:

File: pkg/tarmak/provider/amazon/tf_remote_state.go [hunk 6]
```diff
@@ -385,3 +385,153 @@ func (a *Amazon) initRemoteStateKMS() error {
...
+	svc, err := a.DynamoDB()
+	if err != nil {
+		return err
+	}
+
+	_, err = svc.DeleteItem(&dynamodb.DeleteItemInput{
+		TableName: aws.String(a.RemoteStateName()),
+		Key: map[string]*dynamodb.AttributeValue{
+			"LockID": {
+				S: aws.String(fmt.Sprintf("%s/%s/%s/%s.tfstate-md5", a.RemoteStateName(), a.tarmak.Environment().Name(), a.tarmak.Cluster().Name(), "main")),
```
Comment (JoshVanL) [inline pkg/tarmak/provider/amazon/tf_remote_state.go]: Maybe make this Sprintf into a function

File: pkg/terraform/terraform.go [hunk 7]
```diff
@@ -5,6 +5,7 @@ import (
 	"bufio"
 	"bytes"
 	"fmt"
+	"github.com/jetstack/tarmak/pkg/terraform/providers/tarmak/rpc"
```
Comment (JoshVanL) [inline pkg/terraform/terraform.go]: :sob:

File: pkg/tarmak/tarmak.go [hunk 8]
```diff
@@ -365,3 +374,135 @@ func (t *Tarmak) CmdKubectl(args []string) error {
 func (t *Tarmak) CancellationContext() interfaces.CancellationContext {
 	return t.ctx
 }
+
+func (t *Tarmak) DestroyEnvironment() error {
+	inputDestroy := input.New(os.Stdin, os.Stdout)
+
+	destroyClusters := true
```
Comment (JoshVanL) [inline pkg/tarmak/tarmak.go]: Perhaps create a `var err error` here so we don't have to create lots of response bool variables, i.e
```
var err error
destroy := true
...
destroy, err = inputDestroy(....
...
if destroy {...
.....
```

File: pkg/tarmak/tarmak.go [hunk 9]
```diff
@@ -365,3 +374,135 @@ func (t *Tarmak) CmdKubectl(args []string) error {
...
+		for _, cluster := range t.Environment().Clusters() {
+			hosts, err := cluster.ListHosts()
+			if err != nil {
+				return err
+			}
+			if len(hosts) > 0 {
+				return fmt.Errorf("can't proceed with destroying this environment, because it still has hosts running")
+			}
+		}
+	}
```
Comment (JoshVanL) [inline pkg/tarmak/tarmak.go]: Though I like that it is asking at every step if the user wishes to proceed, I would expect that a no answer would immediately return

File: pkg/tarmak/tarmak.go [hunk 10]
```diff
@@ -365,3 +374,135 @@ func (t *Tarmak) CmdKubectl(args []string) error {
...
+		removeConfig = r
+	}
+
+	if removeConfig {
+		t.log.Infof("Removing environment %v from tarmak.yaml", t.Environment().Name())
+
+		if err := t.config.RemoveEnvironment(t.Environment().Name()); err != nil {
+			return err
+		}
+	}
```
Comment (JoshVanL) [inline pkg/tarmak/tarmak.go]: The current cluster and environment need to be unset or set to something else

File: pkg/tarmak/tarmak.go [hunk 11]
```diff
@@ -365,3 +374,135 @@ func (t *Tarmak) CmdKubectl(args []string) error {
...
+			Query:   "Remove environment from tarmak.yaml?",
+		})
+		if err != nil {
+			return err
+		}
+		removeConfig = r
+	}
+
+	if removeConfig {
+		t.log.Infof("Removing environment %v from tarmak.yaml", t.Environment().Name())
```
Comment (JoshVanL) [inline pkg/tarmak/tarmak.go]: Maybe use all lower case on all 'infof's to match what we are doing elsewhere

File: pkg/tarmak/tarmak.go [hunk 12]
```diff
@@ -365,3 +374,135 @@ func (t *Tarmak) CmdKubectl(args []string) error {
...
+	inputDestroy := input.New(os.Stdin, os.Stdout)
+
+	destroyClusters := true
+	moveFolder := true
+	removeConfig := true
+
+	if !t.flags.Environment.Destroy.AutoApprove {
+		d, err := inputDestroy.AskYesNo(&input.AskYesNo{
+			Default: false,
+			Query:   "Destroy all clusters?",
```
Comment (JoshVanL) [inline pkg/tarmak/tarmak.go]: Maybe print out the environment name again here to triple check the user `Destroy all clusters in environment env?`

File: pkg/tarmak/provider/amazon/tf_remote_state.go [hunk 13]
```diff
@@ -385,3 +385,175 @@ func (a *Amazon) initRemoteStateKMS() error {
...
+
+	result, err := svc.Scan(&dynamodb.ScanInput{
+		TableName: aws.String(a.RemoteStateName()),
+	})
+
+	if err != nil {
+		return 1, err
+	}
+
+	return *result.Count, nil
```
Comment (JoshVanL) [inline pkg/tarmak/provider/amazon/tf_remote_state.go]: Is this ever nil?

File: pkg/tarmak/provider/amazon/tf_remote_state.go [hunk 14]
```diff
@@ -385,3 +385,175 @@ func (a *Amazon) initRemoteStateKMS() error {
...
+	if err != nil {
+		ec2err, ok := err.(awserr.Error)
+		if ok && ec2err.Code() == "BucketNotEmpty" {
+			resp, err := a.listObjectVersionsRemoteState()
+
+			if err != nil {
+				return fmt.Errorf("error listing objects to delete: %s", err)
+			}
+
+			objectsToDelete := a.objectsToDeleteFromRemoteState(resp)
```
Comment (JoshVanL) [inline pkg/tarmak/provider/amazon/tf_remote_state.go]: variable needed here?

File: pkg/tarmak/provider/amazon/tf_remote_state.go [hunk 15]
```diff
@@ -385,3 +385,175 @@ func (a *Amazon) initRemoteStateKMS() error {
...
+
+			params := &s3.DeleteObjectsInput{
+				Bucket: aws.String(a.RemoteStateName()),
+				Delete: &s3.Delete{
+					Objects: objectsToDelete,
+				},
+			}
+
+			_, err = svc.DeleteObjects(params)
+
```
Comment (JoshVanL) [inline pkg/tarmak/provider/amazon/tf_remote_state.go]: newline?

File: pkg/tarmak/provider/amazon/tf_remote_state.go [hunk 16]
```diff
@@ -385,3 +385,175 @@ func (a *Amazon) initRemoteStateKMS() error {
...
+	_, err = svc.DeleteBucket(&s3.DeleteBucketInput{
+		Bucket: aws.String(a.RemoteStateName()),
+	})
+
+	if err != nil {
+		ec2err, ok := err.(awserr.Error)
+		if ok && ec2err.Code() == "BucketNotEmpty" {
+			resp, err := a.listObjectVersionsRemoteState()
+
```
Comment (JoshVanL) [inline pkg/tarmak/provider/amazon/tf_remote_state.go]: new line?

File: pkg/tarmak/tarmak.go [hunk 17]
```diff
@@ -357,3 +366,134 @@ func (t *Tarmak) CmdKubectl(args []string) error {
...
+	}
+
+	return nil
+
+}
+
+func (t *Tarmak) DestroyActivecluster() error {
+	var args []string
+
+	t.log.Infof("Destroying cluster %v", t.Cluster().Name())
```
Comment (JoshVanL) [inline pkg/tarmak/tarmak.go]: Some upper cases still

File: pkg/tarmak/tarmak.go [hunk 18]
```diff
@@ -357,3 +366,134 @@ func (t *Tarmak) CmdKubectl(args []string) error {
...
+		if err = t.config.RemoveEnvironment(t.Environment().Name()); err != nil {
+			return err
+		}
+	}
+
+	return nil
+
+}
+
+func (t *Tarmak) DestroyActivecluster() error {
```
Comment (JoshVanL) [inline pkg/tarmak/tarmak.go]: DestroyActivecuster -> DestroyActiveCluster

File: pkg/tarmak/tarmak.go [hunk 19]
```diff
@@ -357,3 +366,134 @@ func (t *Tarmak) CmdKubectl(args []string) error {
...
+	var args []string
```
Comment (JoshVanL) [inline pkg/tarmak/tarmak.go]: variable needed?

File: pkg/tarmak/tarmak.go [hunk 20]
```diff
@@ -357,3 +366,134 @@ func (t *Tarmak) CmdKubectl(args []string) error {
...
+
+	t.log.Infof("Destroying cluster %v", t.Cluster().Name())
+	destroyCmd := t.NewCmdTerraform(args)
```
Comment (JoshVanL) [inline pkg/tarmak/tarmak.go]: We are not passing signals properly here. The destroy was not respecting my interrupts

File: pkg/tarmak/tarmak.go [hunk 21]
```diff
@@ -357,3 +366,134 @@ func (t *Tarmak) CmdKubectl(args []string) error {
...
+	if err := destroyCmd.Destroy(); err != nil {
+		return err
+	}
+
+	t.log.Infof("Removing S3 state and dynamoDB lock for %v", t.Cluster().Name())
+	if err := t.environment.Provider().Remove(); err != nil {
+		return err
+	}
```
Comment (JoshVanL) [inline pkg/tarmak/tarmak.go]: We should probably be checking context here for signals

File: pkg/tarmak/tarmak.go [hunk 22]
```diff
@@ -357,3 +366,134 @@ func (t *Tarmak) CmdKubectl(args []string) error {
...
 	return t.ctx
 }
+
+func (t *Tarmak) DestroyEnvironment() error {
+	inputDestroy := input.New(os.Stdin, os.Stdout)
+
+	var err error
+
+	destroyClusters := true
+	moveFolder := true
```
Comment (JoshVanL) [inline pkg/tarmak/tarmak.go]: We can use one response variable here

File: pkg/tarmak/tarmak.go [hunk 23]
```diff
@@ -357,3 +366,134 @@ func (t *Tarmak) CmdKubectl(args []string) error {
...
+		archivePath := filepath.Join(t.ConfigPath(), ".archive")
+		environmentArchivePath := filepath.Join(archivePath, t.Environment().Name())
+
+		if _, err = os.Stat(archivePath); err != nil {
+			if os.IsNotExist(err) {
+				if err = os.MkdirAll(archivePath, os.ModePerm); err != nil {
+					return err
+				}
+			} else {
+				return fmt.Errorf("uncatched error: %v", err)
```
Comment (JoshVanL) [inline pkg/tarmak/tarmak.go]: uncatched?

File: pkg/tarmak/tarmak.go [hunk 24]
```diff
@@ -357,3 +366,134 @@ func (t *Tarmak) CmdKubectl(args []string) error {
...
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	if moveFolder {
+		t.log.Info("moving environment folder to .archive")
+
+		archivePath := filepath.Join(t.ConfigPath(), ".archive")
```
Comment (JoshVanL) [inline pkg/tarmak/tarmak.go]: The path would be nice to set as a flag

File: pkg/tarmak/tarmak.go [hunk 25]
```diff
@@ -357,3 +366,134 @@ func (t *Tarmak) CmdKubectl(args []string) error {
...
+				}
+			}
+		}
+
+		// After destroying all other clusters, we can destroy the hub
+		t.cluster = t.Environment().Hub()
+		if err = t.DestroyActivecluster(); err != nil {
+			return err
+		}
+	} else {
```
Comment (JoshVanL) [inline pkg/tarmak/tarmak.go]: I still think that a no here should immediately exit

File: pkg/tarmak/tarmak.go [hunk 26]
```diff
@@ -141,24 +142,33 @@ func (t *Tarmak) initializeModules() {
...
+	t.environment, err = t.EnvironmentByName(currentEnvironmentName)
 	if err != nil {
-		return fmt.Errorf("failed to retrieve current cluster name: %s", err)
+		return err
 	}
-	// init cluster
-	t.cluster, err = t.environment.Cluster(clusterName)
-	if err != nil {
-		return fmt.Errorf("error finding current cluster '%s': %s", clusterName, err)
+	if !environmentDestroy {
```
Comment (JoshVanL) [inline pkg/tarmak/tarmak.go]: Why is this needed?

File: pkg/tarmak/config/config.go [hunk 27]
```diff
@@ -281,6 +281,38 @@ func (c *Config) AppendEnvironment(env *tarmakv1alpha1.Environment) error {
...
+	}
+
+	// Set currentCluster to another cluster if it is part of the environment we want to delete
+	currentConfigEnvironment, err := c.CurrentEnvironmentName()
+	if err != nil {
+		return err
+	}
+	// Pick the first found environment/cluster to replace currentCluster
+	if currentConfigEnvironment == environment {
+		for _, cluster := range c.conf.Clusters {
```
Comment (JoshVanL) [inline pkg/tarmak/config/config.go]: Why can't we just use `	c.SetCurrentCluster(c.tarmak.Clusters()[0].ClusterName())` ?

File: pkg/tarmak/provider/amazon/tf_remote_state.go (hunk 13 above)
Comment (MattiasGees) [reply pkg/tarmak/provider/amazon/tf_remote_state.go]: an `int64` can't be nil.

File: pkg/tarmak/tarmak.go (hunk 26 above)
Comment (MattiasGees) [reply pkg/tarmak/tarmak.go]: Because we give an environment as input, we don't really set an cluster. It tried to pick the cluster from the config, which can fail if the selected cluster is not part of the environment was have as user input.

Going to think a bit more about it on how to solve this in a nicer way. Suggestions are welcome.

File: pkg/tarmak/tarmak.go (hunk 23 above)
Comment (MattiasGees) [reply pkg/tarmak/tarmak.go]: Maybe `unknown` is better?

File: pkg/tarmak/tarmak.go (hunk 24 above)
Comment (MattiasGees) [reply pkg/tarmak/tarmak.go]: It already kinda is because we use the configpath. I still like to keep all config together even when it is archived.

File: pkg/tarmak/provider/amazon/tf_remote_state.go (hunk 13 above)
Comment (JoshVanL) [reply pkg/tarmak/provider/amazon/tf_remote_state.go]: result.Count is a pointer to an int64 which you are dereferencing

File: pkg/tarmak/tarmak.go (hunk 23 above)
Comment (JoshVanL) [reply pkg/tarmak/tarmak.go]: What's wring with 'failed to get status of file: %s"?

File: pkg/tarmak/tarmak.go (hunk 24 above)
Comment (JoshVanL) [reply pkg/tarmak/tarmak.go]: Okay. I think we should have some backup naming scheme if the file is already existing then

File: pkg/tarmak/tarmak.go (hunk 26 above)
Comment (JoshVanL) [reply pkg/tarmak/tarmak.go]: I see, that makes sense. Will have a think

File: pkg/tarmak/provider/amazon/tf_remote_state.go [hunk 28]
```diff
@@ -385,3 +385,175 @@ func (a *Amazon) initRemoteStateKMS() error {
...
+		for _, v := range resp.DeleteMarkers {
+			objectsToDelete = append(objectsToDelete, &s3.ObjectIdentifier{
+				Key:       v.Key,
+				VersionId: v.VersionId,
+			})
+		}
+	}
+
+	if len(resp.Versions) != 0 {
```
[the remark this answers is not in the document]
Comment (MattiasGees) [reply pkg/tarmak/provider/amazon/tf_remote_state.go]: When you delete a object, we also have to delete all versions of it before we can remove a bucket.

File: pkg/tarmak/provider/amazon/tf_remote_state.go [hunk 29]
```diff
@@ -385,3 +385,175 @@ func (a *Amazon) initRemoteStateKMS() error {
...
+		return nil, err
+	}
+	return resp, nil
+}
+
+// Create a list of objects to delete based on a S3 list object
+func (a *Amazon) objectsToDeleteFromRemoteState(resp *s3.ListObjectVersionsOutput) []*s3.ObjectIdentifier {
+	objectsToDelete := make([]*s3.ObjectIdentifier, 0)
+
+	if len(resp.DeleteMarkers) != 0 {
```
[the remark this answers is not in the document]
Comment (MattiasGees) [reply pkg/tarmak/provider/amazon/tf_remote_state.go]: Every deleted object in a versioned bucket has a deleteMarker as it is not fully deleted. In the docs to delete a bucket, it says that we need to be sure that all are gone. https://docs.aws.amazon.com/sdk-for-go/api/service/s3/#S3.DeleteBucket

[the remark this answers is not in the document]
Comment (JoshVanL) [reply pkg/tarmak/provider/amazon/tf_remote_state.go]: Sure but checking the length of a slice before a range is totally unnecessary

File: pkg/tarmak/provider/amazon/tf_remote_state.go (hunk 28 above)
[the remark this answers is not in the document]
Comment (JoshVanL) [reply pkg/tarmak/provider/amazon/tf_remote_state.go]: Also please remove

File: pkg/tarmak/tarmak.go (hunk 20 above)
Comment (MattiasGees) [reply pkg/tarmak/tarmak.go]: Going to use this solution https://github.com/jetstack/tarmak/pull/575#discussion_r225991922

File: cmd/tarmak/cmd/environment_destroy.go [hunk 30]
```diff
@@ -0,0 +1,35 @@
...
+
+// environmentDestroyCmd handles `tarmak environment destroy`
+var environmentDestroyCmd = &cobra.Command{
+	Use:   "destroy [NAME]",
+	Short: "Destroy a environment",
+	PreRunE: func(cmd *cobra.Command, args []string) error {
+		if len(args) == 0 {
+			return errors.New("you have to give the environment name")
+		}
+		globalFlags.Environment.Destroy.Name = args[0]
```
Comment (JoshVanL) [inline cmd/tarmak/cmd/environment_destroy.go]: We shouldn't need to use global flags here, we pass through the args. We should also check == 1 instead of >0

File: pkg/tarmak/cmd.go [hunk 31]
```diff
@@ -103,6 +104,118 @@ func (c *CmdTarmak) Destroy() error {
...
+		// After destroying all other clusters, we can destroy the hub
+		c.cluster = c.Environment().Hub()
+		c.log.Infof("destroying cluster %v", c.Cluster().Name())
+		if err := c.Destroy(); err != nil {
+			return err
+		}
+		if err = c.DestroyClusterMetadata(); err != nil {
+			return err
+		}
+	} else {
```
Comment (JoshVanL) [inline pkg/tarmak/cmd.go]: I still think a no response to deleting all clusters should immediately exit after a no response

File: pkg/tarmak/cmd.go [hunk 32]
```diff
@@ -103,6 +104,118 @@ func (c *CmdTarmak) Destroy() error {
...
+		response, err = inputDestroy.AskYesNo(&input.AskYesNo{
+			Default: false,
+			Query:   "Move environment folder (SSH key and vault_root_token) to .archive?",
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	if response {
```
Comment (JoshVanL) [inline pkg/tarmak/cmd.go]: The logic for these responses is not quite right. You need to at least set response to true again before checking auto-approve

File: pkg/tarmak/cmd.go [hunk 33]
```diff
@@ -103,6 +104,118 @@ func (c *CmdTarmak) Destroy() error {
...
+				if err = os.MkdirAll(archivePath, os.ModePerm); err != nil {
+					return err
+				}
+			} else {
+				return fmt.Errorf("failed to get status of path %v", archivePath)
+			}
+		}
+
+		if _, err = os.Stat(environmentArchivePath); err != nil {
+			if os.IsNotExist(err) {
```
Comment (JoshVanL) [inline pkg/tarmak/cmd.go]: Same here

File: pkg/tarmak/cmd.go [hunk 34]
```diff
@@ -103,6 +104,118 @@ func (c *CmdTarmak) Destroy() error {
...
+				if err = os.Rename(c.Environment().ConfigPath(), environmentArchivePath); err != nil {
+					return err
+				}
+			} else {
+				return fmt.Errorf("failed to get status of path %v", environmentArchivePath)
+			}
+		} else {
```
Comment (JoshVanL) [inline pkg/tarmak/cmd.go]: Same here

File: pkg/tarmak/config/config.go (hunk 27 above)
Comment (JoshVanL) [reply pkg/tarmak/config/config.go]: Can we have a function that lists all clusters in the config?

File: pkg/tarmak/tarmak.go [hunk 35]
```diff
@@ -388,3 +392,17 @@ func (t *Tarmak) verifyImageExists() error {
 
 	return nil
 }
+
+func (t *Tarmak) DestroyClusterMetadata() error {
+	t.log.Infof("removing S3 state and dynamoDB lock for %v", t.Cluster().Name())
```
Comment (JoshVanL) [inline pkg/tarmak/tarmak.go]: This is provider specific, something like `removing tarmak remote state data`

File: pkg/tarmak/tarmak.go [hunk 36]
```diff
@@ -144,24 +144,33 @@ func (t *Tarmak) initializeModules() {
 func (t *Tarmak) initializeConfig() error {
 	var err error
 
+	environmentDestroy := false
+	if t.flags.Environment.Destroy.Name != "" {
```
Comment (JoshVanL) [inline pkg/tarmak/tarmak.go]: We should have a better way to initialize the config. I think we should set the current env to the one to destroy then run this func again

File: pkg/tarmak/provider/amazon/tf_remote_state.go [hunk 37]
```diff
@@ -385,3 +385,169 @@ func (a *Amazon) initRemoteStateKMS() error {
...
+	svc, err := a.S3()
+	if err != nil {
+		return err
+	}
+
+	_, err = svc.DeleteBucket(&s3.DeleteBucketInput{
+		Bucket: aws.String(a.RemoteStateName()),
+	})
+
+	if err != nil {
```
Comment (JoshVanL) [inline pkg/tarmak/provider/amazon/tf_remote_state.go]: if err == nil {
  return
}

helps with indentation

File: pkg/tarmak/provider/amazon/tf_remote_state.go [hunk 38]
```diff
@@ -385,3 +385,169 @@ func (a *Amazon) initRemoteStateKMS() error {
...
+		ec2err, ok := err.(awserr.Error)
+		if ok && ec2err.Code() == "BucketNotEmpty" {
```
Comment (JoshVanL) [inline pkg/tarmak/provider/amazon/tf_remote_state.go]: if !ok || ec2err.Code() != "BucketNotEmpty" {
  return err
}

for indentation again

File: pkg/tarmak/provider/amazon/tf_remote_state.go [hunk 39]
```diff
@@ -385,3 +385,169 @@ func (a *Amazon) initRemoteStateKMS() error {
...
+	}
+
+	result, err := svc.ListObjects(&s3.ListObjectsInput{
+		Bucket: aws.String(a.RemoteStateName()),
+	})
+	if err != nil {
+		return false, err
+	}
+
+	if len(result.Contents) != 0 {
```
Comment (JoshVanL) [inline pkg/tarmak/provider/amazon/tf_remote_state.go]: return len(result.Contents) == 0, nil

File: cmd/tarmak/cmd/environment_destroy.go [hunk 40]
```diff
@@ -0,0 +1,35 @@
...
+import (
+	"errors"
+
+	"github.com/jetstack/tarmak/pkg/tarmak"
+	"github.com/spf13/cobra"
+)
+
+// environmentDestroyCmd handles `tarmak environment destroy`
+var environmentDestroyCmd = &cobra.Command{
+	Use:   "destroy [NAME]",
```
Comment (JoshVanL) [inline cmd/tarmak/cmd/environment_destroy.go]: lowercase name

File: cmd/tarmak/cmd/environment_destroy.go [hunk 41]
```diff
@@ -0,0 +1,35 @@
...
+	Short: "Destroy a environment",
```
Comment (JoshVanL) [inline cmd/tarmak/cmd/environment_destroy.go]: an environment

File: pkg/tarmak/cmd.go [hunk 42]
```diff
@@ -103,6 +104,118 @@ func (c *CmdTarmak) Destroy() error {
...
+			return err
+		}
+	}
+
+	if response {
+		c.log.Info("destroying clusters")
+		for _, cluster := range c.Environment().Clusters() {
+			// We first want to destroy the k8s clusters before the hub
+			if cluster.Name() != c.Environment().Hub().Name() {
+				c.cluster = cluster
```
Comment (JoshVanL) [inline pkg/tarmak/cmd.go]: We should definitely be checking signals around here

**Comment by JoshVanL** [comment]
@MattiasGees  https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

File: pkg/tarmak/cmd.go (hunk 32 above)
Comment (JoshVanL) [reply pkg/tarmak/cmd.go]: Yes sorry, you are quite right

File: pkg/tarmak/tarmak.go (hunk 36 above)
Comment (JoshVanL) [reply pkg/tarmak/tarmak.go]: I don't really like the fact that we are using Name here in our API which is not being used on the CLI. I think it should be possible to set the current environment, `t.config.CurrentCluster = fmt.Sprintf("%s-", args[0])`
then call `t.initializeConfig()` again within `cmd.DestroyEnvrionment()`

**Comment by MattiasGees** [comment]
@JoshVanL Did a bit of digging into https://github.com/jetstack/tarmak/pull/527#pullrequestreview-172098986

This is because the input sends an error on interruption. https://github.com/jetstack/tarmak/blob/b3c28dfadd0b59bdc8d6329a8eb3ca

If needed I can catch it in the input and return something else, but not sure if it is needed.

**Comment by MattiasGees** [comment]
All fixed.

/unassign
/assign @JoshVanL

File: pkg/tarmak/config/config.go [hunk 43]
```diff
@@ -212,6 +212,13 @@ func (c *Config) Environments() (environments []*tarmakv1alpha1.Environment) {
 	return environments
 }
 
+func (c *Config) AllClusters(environment string) (clusters []*clusterv1alpha1.Cluster) {
```
Comment (JoshVanL) [inline pkg/tarmak/config/config.go]: Is this used anywhere?

File: pkg/tarmak/config/config.go [hunk 44]
```diff
@@ -281,6 +288,33 @@ func (c *Config) AppendEnvironment(env *tarmakv1alpha1.Environment) error {
...
+	}
+
+	// Set currentCluster to another cluster if it is part of the environment we want to delete
+	currentConfigEnvironment, err := c.CurrentEnvironmentName()
+	if err != nil {
+		return err
+	}
+	// Pick the first found environment/cluster to replace currentCluster
+	if currentConfigEnvironment == environment {
+		c.SetCurrentCluster(fmt.Sprintf("%v-%v", c.conf.Clusters[0].Environment, c.conf.Clusters[0].Name))
```
Comment (JoshVanL) [inline pkg/tarmak/config/config.go]: We should probably do a range check (>0) on Clusters here

**Comment by MattiasGees** [comment]
/unassign @simonswine 
/assign

Taking this back. Found a bug when destroying a multicluster. In the hub, it doesn't set the socket path for the Tarmak provider.

```
provider "tarmak" {
  socket_path = ""
}
```

**Comment by MattiasGees** [comment]
Fixed it
/unassign
/assign @simonswine
